File: vertex/shared/print-recipe/print-recipe.ts
// Part of the Chili3d Project, under the AGPL-3.0 License.
// See LICENSE file in the project root for full license information.

import { z } from "zod";
import hardnessTable from "./hardness-to-infill-pct.json";
/** Locked clinical hardness ladder (Biomechanics 2026-09-20). */
export const HARDNESS_NAMES = ["Extra Soft", "Soft", "Medium", "Hard", "Extra Hard"] as const;

export type HardnessName = (typeof HARDNESS_NAMES)[number];

export type GyroidInfillPct = 14 | 20 | 26 | 32 | 38;

export const HARDNESS_TO_INFILL_PCT: Record<HardnessName, GyroidInfillPct> = hardnessTable as Record<
    HardnessName,
    GyroidInfillPct
>;

const gyroidPctSchema = z.union([z.literal(14), z.literal(20), z.literal(26), z.literal(32), z.literal(38)]);

const hardnessNameSchema = z.enum(HARDNESS_NAMES);

/** Phase A production default — belt TPU profile (Track 0 migration target). */
export const DEFAULT_PRODUCTION_PROFILE_ID = "apex-belt-v2";

export const printRecipeV1Schema = z.object({
    version: z.literal(1),
    pattern: z.literal("gyroid"),
    /** Locked printer/material profile (Phase A: existing PRINTER_PRESETS id). */
    profileId: z.string().min(1),
    defaultHardness: hardnessNameSchema,
    zones: z.tuple([]),
    hardnessToInfillPct: z.record(hardnessNameSchema, gyroidPctSchema).optional(),
    /** Belt release label in manufacturing output (default ON). */
    includeProductionLabel: z.boolean().optional(),
});

export type PrintRecipeV1 = z.infer<typeof printRecipeV1Schema>;

export const DEFAULT_PRINT_RECIPE_V1: PrintRecipeV1 = {
    version: 1,
    pattern: "gyroid",
    profileId: DEFAULT_PRODUCTION_PROFILE_ID,
    defaultHardness: "Medium",
    zones: [],
    includeProductionLabel: true,
};

/** Recommended default for belt production release labels (Track 3a). */
export const DEFAULT_INCLUDE_PRODUCTION_LABEL = true;

/** Presets that require PrintRecipeV1 (reject legacy infill-only overrides). */
export const GYROID_MANUFACTURING_PRESET_IDS = new Set([
    "apex-belt-v2",
    "apex-belt-v2-shell",
    "desktop-fdm",
    "apex-belt-v2-45",
    "layerloop-30",
]);

export function isGyroidManufacturingPreset(presetId: string): boolean {
    return GYROID_MANUFACTURING_PRESET_IDS.has(presetId);
}

export function gyroidInfillPctForHardness(
    hardness: HardnessName,
    table: Partial<Record<HardnessName, GyroidInfillPct>> | undefined = HARDNESS_TO_INFILL_PCT,
): GyroidInfillPct {
    const pct = table?.[hardness] ?? HARDNESS_TO_INFILL_PCT[hardness];
    return pct;
}

/** Volume fraction 0..1 for slicer APIs that use scalar density. */
export function infillFractionFromRecipe(recipe: PrintRecipeV1): number {
    const pct = gyroidInfillPctForHardness(recipe.defaultHardness, recipe.hardnessToInfillPct);
    return pct / 100;
}

function withLabelDefault(recipe: PrintRecipeV1): PrintRecipeV1 {
    return {
        ...recipe,
        includeProductionLabel: recipe.includeProductionLabel ?? true,
    };
}

/** Legacy PrintRecipeV1 payloads saved before profile lock (no profileId). */
const legacyPrintRecipeV1Schema = z.object({
    version: z.literal(1),
    pattern: z.literal("gyroid"),
    defaultHardness: hardnessNameSchema,
    zones: z.tuple([]),
    hardnessToInfillPct: z.record(hardnessNameSchema, gyroidPctSchema).optional(),
    includeProductionLabel: z.boolean().optional(),
});

export function migratePrintRecipe(recipe: PrintRecipeV1 | undefined | null): PrintRecipeV1 {
    if (!recipe) {
        return { ...DEFAULT_PRINT_RECIPE_V1 };
    }
    const parsed = printRecipeV1Schema.safeParse(recipe);
    if (parsed.success) {
        return withLabelDefault(parsed.data);
    }
    const legacy = legacyPrintRecipeV1Schema.safeParse(recipe);
    if (legacy.success) {
        return withLabelDefault({
            ...legacy.data,
            profileId: DEFAULT_PRODUCTION_PROFILE_ID,
        });
    }
    return { ...DEFAULT_PRINT_RECIPE_V1 };
}

export class PrintRecipeProfileRequiredError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "PrintRecipeProfileRequiredError";
    }
}

/**
 * Gyroid / parity manufacturing requires an explicit PrintRecipe with profileId.
 * Legacy designs are upgraded via migratePrintRecipe when profileId was omitted on save.
 */
export function resolveGyroidManufacturingPrintRecipe(
    presetId: string,
    printRecipeInput: PrintRecipeV1 | undefined,
): PrintRecipeV1 {
    if (!isGyroidManufacturingPreset(presetId)) {
        throw new PrintRecipeProfileRequiredError(
            "resolveGyroidManufacturingPrintRecipe called for non-gyroid preset",
        );
    }
    if (!printRecipeInput) {
        throw new PrintRecipeProfileRequiredError(
            "Production gyroid manufacturing requires PrintRecipeV1 with profileId (locked printer preset).",
        );
    }
    const recipe = migratePrintRecipe(printRecipeInput);
    if (!recipe.profileId?.trim()) {
        throw new PrintRecipeProfileRequiredError(
            "PrintRecipeV1.profileId is required for gyroid manufacturing.",
        );
    }
    return recipe;
}

/** Snake_case payload for Python `/manufacture`. */
export function printRecipeToSnake(recipe: PrintRecipeV1): Record<string, unknown> {
    const migrated = migratePrintRecipe(recipe);
    return {
        version: migrated.version,
        pattern: migrated.pattern,
        profile_id: migrated.profileId,
        default_hardness: migrated.defaultHardness,
        zones: migrated.zones,
        hardness_to_infill_pct: migrated.hardnessToInfillPct ?? HARDNESS_TO_INFILL_PCT,
        include_production_label: migrated.includeProductionLabel ?? true,
    };
}

/** Bind hardness + gyroid ladder to a locked production profile (preset id). */
export function bindPrintRecipeProfile(
    recipe: PrintRecipeV1 | undefined | null,
    profileId: string,
    hardness?: HardnessName,
): PrintRecipeV1 {
    const base = migratePrintRecipe(recipe);
    return withLabelDefault({
        ...base,
        profileId,
        defaultHardness: hardness ?? base.defaultHardness,
    });
}

export function setProductionLabelEnabled(
    recipe: PrintRecipeV1 | undefined | null,
    enabled: boolean,
): PrintRecipeV1 {
    const base = migratePrintRecipe(recipe);
    return { ...base, includeProductionLabel: enabled };
}

File: python-service/app/services/print_recipe.py
# Part of the Chili3d Project, under the AGPL-3.0 License.
# See LICENSE file in the project root for full license information.

"""
PrintRecipeV1 resolution for hybrid manufacturing (Phase A).

Shares the locked hardness→% ladder with the TypeScript client via
vertex/shared/print-recipe/hardness-to-infill-pct.json.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Literal

from pydantic import BaseModel, Field

HardnessName = Literal["Extra Soft", "Soft", "Medium", "Hard", "Extra Hard"]

_TABLE_PATH = (
    Path(__file__).resolve().parents[3]
    / "vertex"
    / "shared"
    / "print-recipe"
    / "hardness-to-infill-pct.json"
)


def load_hardness_table() -> dict[str, int]:
    with _TABLE_PATH.open(encoding="utf-8") as f:
        data = json.load(f)
    return {str(k): int(v) for k, v in data.items()}


HARDNESS_TO_INFILL_PCT: dict[str, int] = load_hardness_table()

DEFAULT_PRODUCTION_PROFILE_ID = "apex-belt-v2"

GYROID_MANUFACTURING_PRESET_IDS = frozenset(
    {
        "apex-belt-v2",
        "apex-belt-v2-shell",
        "desktop-fdm",
        "apex-belt-v2-45",
        "layerloop-30",
    }
)


class SoleUvFramePayload(BaseModel):
    min_x_mm: float
    min_y_mm: float
    length_mm: float
    width_mm: float


class PrintRecipeV1(BaseModel):
    version: Literal[1] = 1
    pattern: Literal["gyroid"] = "gyroid"
    profile_id: str = Field(default=DEFAULT_PRODUCTION_PROFILE_ID, min_length=1)
    default_hardness: HardnessName = "Medium"
    zones: list[Any] = Field(default_factory=list)
    hardness_to_infill_pct: dict[str, int] | None = None
    override_soft_wins: bool = False
    sole_uv_frame: SoleUvFramePayload | None = None

    def resolved_infill_percent(self) -> int:
        table = self.hardness_to_infill_pct or HARDNESS_TO_INFILL_PCT
        pct = table.get(self.default_hardness)
        if pct is None:
            pct = HARDNESS_TO_INFILL_PCT[self.default_hardness]
        return int(pct)

    def infill_fraction(self) -> float:
        return self.resolved_infill_percent() / 100.0


DEFAULT_PRINT_RECIPE = PrintRecipeV1()


def is_gyroid_manufacturing_preset(preset_id: str) -> bool:
    return preset_id in GYROID_MANUFACTURING_PRESET_IDS


def coerce_print_recipe(raw: PrintRecipeV1 | dict[str, Any] | None) -> PrintRecipeV1:
    if raw is None:
        return DEFAULT_PRINT_RECIPE.model_copy()
    if isinstance(raw, PrintRecipeV1):
        return raw
    data = dict(raw)
    if not str(data.get("profile_id") or "").strip():
        data["profile_id"] = DEFAULT_PRODUCTION_PROFILE_ID
    return PrintRecipeV1.model_validate(data)


def require_gyroid_print_recipe_with_profile(
    preset_id: str, print_recipe: PrintRecipeV1 | None
) -> PrintRecipeV1:
    if not is_gyroid_manufacturing_preset(preset_id):
        raise ValueError("require_gyroid_print_recipe_with_profile called for non-gyroid preset")
    if print_recipe is None:
        raise ValueError(
            "Production gyroid manufacturing requires print_recipe with profile_id (locked printer preset)."
        )
    recipe = coerce_print_recipe(print_recipe)
    if not recipe.profile_id.strip():
        raise ValueError("print_recipe.profile_id is required for gyroid manufacturing.")
    if recipe.profile_id != preset_id:
        raise ValueError("preset_id must match print_recipe.profile_id for production gyroid manufacturing.")
    return recipe
